File: test/pathutil.test.ts
import assert from "node:assert/strict";
import { describe, it } from "node:test";
import { PathError } from "../src/errors.js";
import { normalizeRelPath, uniqueNeededPaths } from "../src/pathutil.js";

describe("pathutil", () => {
  it("converts backslashes, NFC, and matches server fileset rules", () => {
    const decomposed = "cafe\u0301/file.bin";
    const composed = "café/file.bin";
    assert.equal(normalizeRelPath(decomposed), normalizeRelPath(composed));
    assert.equal(normalizeRelPath("a\\b\\c"), "a/b/c");
    assert.equal(normalizeRelPath("bin//sub\\\\dir///file.txt"), "bin/sub/dir/file.txt");
    assert.equal(normalizeRelPath("folder/subfolder/"), "folder/subfolder");
    assert.throws(() => normalizeRelPath("../secret"), PathError);
    assert.throws(() => normalizeRelPath("ok/../../x"), PathError);
    assert.throws(() => normalizeRelPath("./file.txt"), PathError);
    assert.throws(() => normalizeRelPath("dir/./file.txt"), PathError);
    assert.throws(() => normalizeRelPath("/etc/passwd"), PathError);
    assert.throws(() => normalizeRelPath("C:/Windows/app.exe"), PathError);
    assert.throws(() => normalizeRelPath("foo/C:/bar"), PathError);
    assert.throws(() => normalizeRelPath("dir/file\x00.txt"), PathError);
    assert.throws(() => normalizeRelPath(""), PathError);
  });

  it("uniqueNeededPaths is order-stable unique NFC", () => {
    const a = uniqueNeededPaths(["b.bin", "a.bin", "a.bin", "b.bin"]);
    const b = uniqueNeededPaths(["a.bin", "b.bin"]);
    assert.deepEqual(a, b);
  });
});

File: test/pathutil_test.dart
import 'package:kirivers_client/kirivers_client.dart';
import 'package:test/test.dart';

void main() {
  group('normalizePath', () {
    test('converts backslashes and NFC', () {
      expect(normalizePath(r'bin\resources\config.json'),
          'bin/resources/config.json');
      const nfd = 'caf\u0065\u0301.txt';
      const nfc = 'caf\u00e9.txt';
      expect(normalizePath(nfd), normalizePath(nfc));
      expect(
          normalizePath('bin//sub\\\\dir///file.txt'), 'bin/sub/dir/file.txt');
    });

    test('rejects traversal and absolute paths', () {
      expect(() => normalizePath('../x'), throwsA(isA<PathException>()));
      expect(() => normalizePath('/etc/passwd'), throwsA(isA<PathException>()));
      expect(() => normalizePath(r'C:\Windows'), throwsA(isA<PathException>()));
      expect(
          () => normalizePath('dir/./file.txt'), throwsA(isA<PathException>()));
      expect(() => normalizePath(''), throwsA(isA<PathException>()));
    });

    test('uniqueNormalizedPaths drops duplicates after NFC', () {
      final got = uniqueNormalizedPaths([
        r'a\b.txt',
        'a/b.txt',
        'caf\u0065\u0301.txt',
        'caf\u00e9.txt',
      ]);
      expect(got, ['a/b.txt', 'caf\u00e9.txt']);
    });
  });
}
